pyclibrary/c_parser.py

```diff
@@ -31,6 +31,215 @@
+                 extra_type_list)
+        for w in self[0].split():
+            if w not in names:
+                return False
+        return True
+
+    def eval(self, type_map, used=None):
+        """Resolves the type_spec of this type recursively if it is referring
+        to a typedef. For resolving the type type_map is used for lookup.
+        Returns a new Type object.
+
+        Parameters
+        ----------
+        type_map : dict of str to Type
+            All typedefs that shall be resolved have to be stored in this
+            type_map.
+
+        used : list of str
+            For internal use only to prevent circular typedefs
+
+        """
+        used = used or []
+
+        if self.is_fund_type():
+            # Remove 'signed' before returning evaluated type
+            return Type(re.sub(r'\bsigned\b', '', self.type_spec).strip(),
+                        *self.declarators)
+
+        parent = self.type_spec
+        if parent in used:
+            m = 'Recursive loop while evaluating types. (typedefs are {})'
+            raise DefinitionError(m.format(' -> '.join(used+[parent])))
+
+        used.append(parent)
+        if parent not in type_map:
+            m = 'Unknown type "{}" (typedefs are {})'
+            raise DefinitionError(m.format(parent, ' -> '.join(used)))
+
+        pt = type_map[parent]
+        evaled_type = Type(pt.type_spec, *(pt.declarators + self.declarators))
+        return evaled_type.eval(type_map, used)
+
+    def add_compatibility_hack(self):
+        """If This Type is refering to a function (**not** a function pointer)
+        a new type is returned, that matches the hack from version 0.1.0.
+        This hack enforces the return value be encapsulated in a separated Type
+        object:
+
+            Type('int', '*', ())
+
+        is converted to
+
+            Type(Type('int', '*'), ())
+        """
+        if type(self[-1]) == tuple:
+            return Type(Type(*self[:-1]), self[-1])
+        else:
+            return self
+
+    def remove_compatibility_hack(self):
+        """Returns a Type object, where the hack from .add_compatibility_hack()
+        is removed
+
+        """
+        if len(self) == 2 and isinstance(self[0], Type):
+            return Type(*(self[0] + (self[1],)))
+        else:
+            return self
+
+    def __repr__(self):
+        return type(self).__name__ + '(' + ', '.join(map(repr, self)) + ')'
+
+
+class Compound(dict):
+
+    def __init__(self, *members, **argv):
+        self.members = list(members)
+        self.pack = argv.pop('pack', None)
+        assert len(argv) == 0
+
+        super(Compound, self).__init__(
+            dict(members=self.members, pack=self.pack))
+
+    def __repr__(self):
+        packParam = ', pack='+repr(self.pack) if self.pack is not None else ''
+        return (type(self).__name__ + '(' +
+                ', '.join(map(repr, self.members)) + packParam + ')')
+
+
+class Struct(Compound):
```

Comment:
If we slot Compound you will to declare __slots__ = () in subclasses to avoid them having an instance dict.